src/App.js

```diff
@@ -26,46 +29,80 @@ const generateSquares = () => {
 }
 
 const App = () => {
+  
+    const [squares, setSquares] = useState(generateSquares());
+    const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+    const [numSquaresFilled, setNumSquaresFilled] = useState(0);
+    const [winner, setWinner] = useState(null);
 
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
-  const [squares, setSquares] = useState(generateSquares());
+    let updatedNumSquaresFilled = numSquaresFilled
 
-  // Wave 2
-  // You will need to create a method to change the square 
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+    // Wave 2: Play game
+  
+    const onClickCallback = (id) => {
+    if (winner !== null) return;
 
+    const newSquares = [...squares];
 
-  const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
-
+    for (const row of newSquares) {
+      for (const square of row) {
+        if (square.id === id && square.value === '') {
+          square.value = currentPlayer;
+          updatedNumSquaresFilled = numSquaresFilled + 1
+          setNumSquaresFilled(updatedNumSquaresFilled);
+          currentPlayer === PLAYER_1 ? setCurrentPlayer(PLAYER_2) : setCurrentPlayer(PLAYER_1);
+        }
+      }
+    }
+    checkForWinner();
+    setSquares(newSquares);
   }
 
-  const resetGame = () => {
-    // Complete in Wave 4
-  }
+    // Wave 3: Check for winner
 
-  return (
-    <div className="App">
-      <header className="App-header">
-        <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
-      </header>
-      <main>
-        <Board squares={squares} />
-      </main>
-    </div>
-  );
-}
+    const checkForWinner = () => {
+
+      const score = squares.flat()
+      const winners = [[0, 1, 2],[3, 4, 5],[6, 7, 8],[0, 3, 6],[1, 4, 7],[2, 5, 8],[0, 4, 8],[2, 4, 6]];
```

Comment:
This is clever. I would consider adding a comment here to describe simply how this algorithm works -- It looks like indices for rows, columns, and diagonals are stored in the winners array. I would also consider how you could implement this with a loop that finds the rows, columns, diagonals in the original 2D squares array.